pdu/Src/cooling.c

```diff
@@ -0,0 +1,93 @@
+#include "cooling.h"
+
+#include <stdbool.h>
+#include "FreeRTOS.h"
+#include "task.h"
+#include "state_machine.h"
+#include "controlStateMachine.h"
+#include "FreeRTOS_CLI.h"
+#include "debug.h"
+#include "pdu_can.h"
+#include "watchdog.h"
+#include "canReceive.h"
+#include "pdu_dtc.h"
+#include "LTC4110.h"
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+#define SCALE_TEMP_PERCENT(temp, max_temp) (((temp) - COOLING_AMBIENT_TEMP_C) / ((max_temp) - COOLING_AMBIENT_TEMP_C))
+
+void coolingOff(void) {
+    DEBUG_PRINT("Turning cooling off\n");
+    PUMP_LEFT_DISABLE;
+    PUMP_RIGHT_DISABLE;
+    FAN_LEFT_DISABLE;
+    FAN_RIGHT_DISABLE;
+}
+
+void coolingOn(void) {
+    DEBUG_PRINT("Turning cooling on\n");
+    PUMP_LEFT_ENABLE;
+    PUMP_RIGHT_ENABLE;
+    FAN_LEFT_ENABLE;
+    FAN_RIGHT_ENABLE;
+}
+
+void coolingTask(void *pvParameters) {
+    if (registerTaskToWatch(COOLING_TASK_ID, 2*COOLING_TASK_PERIOD_MS, false, NULL) != HAL_OK)
+    {
+        ERROR_PRINT("Failed to register cooling task with watchdog!\n");
+        Error_Handler();
+    }
+
+    TickType_t xLastWakeTime = xTaskGetTickCount();
+
+    uint32_t cooling_state = COOL_STATE_WAIT;
+
+    while(1)
+    {
+        // Only cool if motor controller on and EM enabled
+        // Should we check if MC is on?
+        if (fsmGetState(&mainFsmHandle) == MN_STATE_Motors_On && DC_DC_state) {
+            cooling_state = COOL_STATE_ON;
+            DEBUG_PRINT("Cooling enabled\n");
+        } else {
+            cooling_state = COOL_STATE_OFF;
+            DEBUG_PRINT("Cooling disabled\n");
+        }
+
+        if (cooling_state == COOL_STATE_ON)
+        {
+            // Check if we are receiving temp readings
+            if (xTaskGetTickCount() - MC_Last_Temperature_Msg_ticks > pdMS_TO_TICKS(MC_MESSAGE_LOST_THRESHOLD_MS)) {
+                sendDTC_WARNING_MC_Temp_Reading_Lost();
```

Comment:
not sure if this is needed but it would be interesting to know if we do lose values